torch/profiler/profiler.py

```diff
@@ -0,0 +1,192 @@
+            with_stack=False):
+        self.activities = activities
+        self.enable_pred = enable_pred
+        self.record_shapes = record_shapes
+        self.profile_memory = profile_memory
+        self.with_stack = with_stack
+        self.step_num = 0
+        self.profiler: Optional[prof.profile] = None
+        self.step_rec_fn: Optional[prof.record_function] = None
+
+        if not self.enable_pred:
+            print("Warning: using profiler without enable predicate may result in the skewed " +
+                    "results, use enable_pred to control the warmup time")
+
+    def __enter__(self):
+        self.next_step()
+        if not self.enable_pred:
+            self._run_action(EnablePred.Action.START_WARMUP)
+            self._run_action(EnablePred.Action.START_TRACE)
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self.step_rec_fn:
+            self.step_rec_fn.__exit__(None, None, None)
+        if self.profiler:
+            if self.enable_pred:
+                if self.enable_pred._num_state(self.step_num) == EnablePred.State.WARMUP:
+                    self._run_action(EnablePred.Action.START_TRACE)
+            self._run_action(EnablePred.Action.STOP_TRACE, keep_profiler=True)
+
+    def next_step(self):
+        if self.step_rec_fn:
+            self.step_rec_fn.__exit__(None, None, None)
+        self.step_num += 1
+        if self.enable_pred:
+            self._run_actions(self.step_num)
+
+        self.step_rec_fn = prof.record_function("ProfilerStep#" + str(self.step_num))
+        self.step_rec_fn.__enter__()
+
+    def export_chrome_trace(self, path: str):
+        assert self.profiler
+        return self.profiler.export_chrome_trace(path)
+
+    def key_averages(self, group_by_input_shape: bool = False, group_by_stack_n: int = 0):
+        assert self.profiler
+        return self.profiler.key_averages(group_by_input_shape, group_by_stack_n)
+
+    def _run_actions(self, step_num):
+        assert self.enable_pred
+        for act in self.enable_pred.actions(self.step_num):
+            self._run_action(act)
+
+    def _run_action(self, act, keep_profiler=False):
+        if act == EnablePred.Action.START_WARMUP:
+            self.profiler = prof.profile(
+                use_cuda=(ProfilerActivity.CUDA in self.activities),
+                use_cpu=(ProfilerActivity.CPU in self.activities),
+                record_shapes=self.record_shapes,
+                profile_memory=self.profile_memory,
+                with_stack=self.with_stack,
+                use_kineto=True,
+            )
+            self.profiler._prepare_kineto_trace()
+        elif act == EnablePred.Action.START_TRACE:
+            assert self.profiler is not None
+            self.profiler._start_kineto_trace()
+        elif act == EnablePred.Action.STOP_TRACE:
+            assert self.profiler is not None
+            self.profiler.__exit__(None, None, None)
+            assert self.enable_pred
```

Comment:
This assertion will fail in case  self.enable_pred=None

Reply:
right, good catch, thanks!

Reply:
was going to add a non-enable_pred unit test too

Reply:
fixed and added test